
This patch was submitted to Apache:
	https://issues.apache.org/bugzilla/show_bug.cgi?id=44851
	
It adds PUT and DELETE support to the apache utility, 'ApacheBench'.


--- ab.c.original       2008-04-21 08:46:05.000000000 -0700
+++ ab.c        2008-04-22 07:16:39.000000000 -0700
@@ -91,7 +91,7 @@
  * ab - or to due to a change in the distribution it is compiled with
  * (such as an APR change in for example blocking).
  */
-#define AP_AB_BASEREVISION "2.0.40-dev"
+#define AP_AB_BASEREVISION "2.0.41-dev"
 
 /*
  * BUGS:
@@ -254,7 +254,7 @@
 /* --------------------- GLOBALS ---------------------------- */
 
 int verbosity = 0;      /* no verbosity by default */
-int posting = 0;        /* GET by default */
+int method = 0;        /* GET by default */
 int requests = 1;       /* Number of requests to make */
 int heartbeatres = 100; /* How often do we say we're alive */
 int concurrency = 1;    /* Number of multiple requests to make */
@@ -621,7 +621,7 @@
             c->connect = tnow;
             c->rwrite = reqlen;
             c->rwrote = 0;
-            if (posting)
+            if (method)
                 c->rwrite += postlen;
         }
         else if (tnow > c->connect + aprtimeout) {
@@ -761,8 +761,10 @@
     if (keepalive)
         printf("Keep-Alive requests:    %ld\n", doneka);
     printf("Total transferred:      %ld bytes\n", totalread);
-    if (posting > 0)
+    if (method == 1)
         printf("Total POSTed:           %ld\n", totalposted);
+    if (method == 2)
+        printf("Total PUT:              %ld\n", totalposted);
     printf("HTML transferred:       %ld bytes\n", totalbread);
 
     /* avoid divide by zero */
@@ -775,7 +777,7 @@
            (float) (1000 * timetaken / done));
         printf("Transfer rate:          %.2f [Kbytes/sec] received\n",
            (float) (totalread / 1024 / timetaken));
-        if (posting > 0) {
+        if (method > 0) {
             printf("                        %.2f kb/s sent\n",
                (float) (totalposted / timetaken / 1024));
             printf("                        %.2f kb/s total\n",
@@ -1030,10 +1032,14 @@
     printf("<tr %s><th colspan=2 %s>Total transferred:</th>"
        "<td colspan=2 %s>%ld bytes</td></tr>\n",
        trstring, tdstring, tdstring, totalread);
-    if (posting > 0)
+    if (method == 1)
         printf("<tr %s><th colspan=2 %s>Total POSTed:</th>"
            "<td colspan=2 %s>%ld</td></tr>\n",
            trstring, tdstring, tdstring, totalposted);
+    if (method == 2)
+        printf("<tr %s><th colspan=2 %s>Total PUT:</th>"
+           "<td colspan=2 %s>%ld</td></tr>\n",
+           trstring, tdstring, tdstring, totalposted);
     printf("<tr %s><th colspan=2 %s>HTML transferred:</th>"
        "<td colspan=2 %s>%ld bytes</td></tr>\n",
        trstring, tdstring, tdstring, totalbread);
@@ -1046,7 +1052,7 @@
         printf("<tr %s><th colspan=2 %s>Transfer rate:</th>"
            "<td colspan=2 %s>%.2f kb/s received</td></tr>\n",
            trstring, tdstring, tdstring, (float) (totalread) / timetaken);
-        if (posting > 0) {
+        if (method > 0) {
             printf("<tr %s><td colspan=2 %s>&nbsp;</td>"
                "<td colspan=2 %s>%.2f kb/s sent</td></tr>\n",
                trstring, tdstring, tdstring,
@@ -1480,6 +1486,7 @@
 
 static void test(void)
 {
+       char *verb = "GET";
     apr_time_t now;
     apr_int16_t rv;
     long i;
@@ -1543,24 +1550,46 @@
     }
 
     /* setup request */
-    if (posting <= 0) {
+    if (method <= 0) {
+        switch (method) {
+                       case -2:
+                               verb = "DELETE";
+                               break;
+                       case -1:
+                               verb = "HEAD";
+                               break;
+                       case 0:
+                               verb = "GET";
+                               break;
+               }
+
         snprintf_res = apr_snprintf(request, sizeof(_request),
             "%s %s HTTP/1.0\r\n"
             "%s" "%s" "%s"
             "%s" "\r\n",
-            (posting == 0) ? "GET" : "HEAD",
+            verb,
             (isproxy) ? fullurl : path,
             keepalive ? "Connection: Keep-Alive\r\n" : "",
             cookie, auth, hdrs);
     }
     else {
+        switch (method) {
+                       case 1:
+                               verb = "POST";
+                               break;
+                       case 2:
+                               verb = "PUT";
+                               break;
+               }
+
         snprintf_res = apr_snprintf(request,  sizeof(_request),
-            "POST %s HTTP/1.0\r\n"
+            "%s %s HTTP/1.0\r\n"
             "%s" "%s" "%s"
             "Content-length: %" APR_SIZE_T_FMT "\r\n"
             "Content-type: %s\r\n"
             "%s"
             "\r\n",
+                       verb,
             (isproxy) ? fullurl : path,
             keepalive ? "Connection: Keep-Alive\r\n" : "",
             cookie, auth,
@@ -1577,9 +1606,9 @@
     reqlen = strlen(request);
 
     /*
-     * Combine headers and (optional) post file into one contineous buffer
+     * Combine headers and (optional) post file into one continuous buffer
      */
-    if (posting == 1) {
+    if (method >= 1) {
         char *buff = malloc(postlen + reqlen + 1);
         if (!buff) {
             fprintf(stderr, "error creating request buffer: out of memory\n");
@@ -1779,10 +1808,12 @@
     fprintf(stderr, "    -c concurrency  Number of multiple requests to make\n");
     fprintf(stderr, "    -t timelimit    Seconds to max. wait for responses\n");
     fprintf(stderr, "    -p postfile     File containing data to POST\n");
+    fprintf(stderr, "    -u putfile      File containing data to PUT\n");
     fprintf(stderr, "    -T content-type Content-type header for POSTing\n");
     fprintf(stderr, "    -v verbosity    How much troubleshooting info to print\n");
     fprintf(stderr, "    -w              Print out results in HTML tables\n");
-    fprintf(stderr, "    -i              Use HEAD instead of GET\n");
+    fprintf(stderr, "    -i              Send a HEAD request\n");
+    fprintf(stderr, "    -D              Send a DELETE request\n");
     fprintf(stderr, "    -x attributes   String to insert as table attributes\n");
     fprintf(stderr, "    -y attributes   String to insert as tr attributes\n");
     fprintf(stderr, "    -z attributes   String to insert as td or th attributes\n");
@@ -1962,7 +1993,7 @@
 #endif
 
     apr_getopt_init(&opt, cntxt, argc, argv);
-    while ((status = apr_getopt(opt, "n:c:t:T:p:v:kVhwix:y:z:C:H:P:A:g:X:de:Sq"
+    while ((status = apr_getopt(opt, "n:c:t:T:p:v:kVhwix:y:z:C:H:P:A:g:X:de:Sq:Du:"
 #ifdef USE_SSL
             "Z:f:"
 #endif
@@ -1984,9 +2015,12 @@
                 concurrency = atoi(optarg);
                 break;
             case 'i':
-                if (posting == 1)
-                err("Cannot mix POST and HEAD\n");
-                posting = -1;
+                if (method != 0) err("Cannot mix HTTP methods\n");
+                method = -1;
+                break;
+            case 'D':
+                if (method != 0) err("Cannot mix HTTP methods\n");
+                method = -2;
                 break;
             case 'g':
                 gnuplot = strdup(optarg);
@@ -2001,10 +2035,20 @@
                 confidence = 0;
                 break;
             case 'p':
-                if (posting != 0)
-                    err("Cannot mix POST and HEAD\n");
+                if (method != 0)
+                    err("Cannot mix HTTP methods\n");
+                if (0 == (r = open_postfile(optarg))) {
+                    method = 1;
+                }
+                else if (postdata) {
+                    exit(r);
+                }
+                break;
+            case 'u':
+                if (method != 0)
+                    err("Cannot mix HTTP methods\n");
                 if (0 == (r = open_postfile(optarg))) {
-                    posting = 1;
+                    method = 2;
                 }
                 else if (postdata) {
                     exit(r);

